### ISSUE TITLE
unable to load plugin hipSYCL_clang.so: undefined symbol: _ZN4llvm24DisableABIBreakingChecksE

### ISSUE THREAD

**Issue by WageeshaR** [AdaptiveCpp/AdaptiveCpp#743]
I'm trying to compile a simple c++ benchmark using hipsycl with cuda. Apparently an ABI mismatch is happenning when loading the Clang plugin but I have no clue. I rebuilt hipsycl making sure the LLVM/Clang versions are correctly picked up (check below trace) but when using it fails.

This is the trace for configuring the build. It correctly picks up LLVM and Clang.
command:
`cmake -DCMAKE_INSTALL_PREFIX=./ DCUDA_TOOLKIT_ROOT_DIR=/opt/apps/easybuild/software/CUDA/11.3.1 -DWITH_CUDA_BACKEND=ON ../`
```
-- Could not find HIP cmake integration, falling back to finding hipcc
-- Could not find ROCm installation by looking for hipcc. ROCm support unavailable.
-- Found CUDA driver library: /opt/apps/easybuild/software/CUDA/11.3.1/lib64/stubs/libcuda.so
-- Building hipSYCL against LLVM configured from /opt/apps/easybuild/software/LLVM/12.0.1-GCCcore-11.2.0/lib64/cmake/llvm
-- Selecting clang: /opt/apps/easybuild/software/Clang/12.0.1-GCCcore-11.2.0/bin/clang++
-- Using clang include directory: /opt/apps/easybuild/software/Clang/12.0.1-GCCcore-11.2.0/lib/clang/12.0.1/include/..
-- Configuring done
-- Generating done
-- Build files have been written to: /users/wrs505/scratch/indv-project/sycl-workspace/hipSYCL/build
```
This is the Makefile used for the benchmark application.
```
CXX       = syclcc
OPTIMIZE  = yes
DEBUG     = no
GPU       = no
CUDA      = no
CUDA_ARCH = sm_70
HIP       = no
HIP_ARCH  = gfx908
GCC_TOOLCHAIN = "/opt/apps/easybuild/software/GCCcore/11.2.0"

CPPFLAGS = --hipsycl-targets=cuda.integrated-multipass:sm_70 -I. -I../utils -I../fem $(SOME_DIRECTIVE) -I../../include/ $(SOME_OTHER_DIRECTIVE) \
           -std=c++17 -Wall

include make_targets
```
This is the output when trying to `make`
```
CXX: /users/wrs505/scratch/indv-project/sycl-workspace/hipSYCL/bin/syclcc
Compiler version: unknown
syclcc -O3 --hipsycl-targets=cuda.integrated-multipass:sm_70 -I. -I../utils -I../fem -DMINIFE_SCALAR=double -DMINIFE_LOCAL_ORDINAL=int -DMINIFE_GLOBAL_ORDINAL=int -DMINIFE_RESTRICT=__restrict__ -I../../include/ -DMINIFE_CSR_MATRIX -std=c++17 -Wall -DMINIFE_INFO=1 -DMINIFE_KERNELS=0 -c main.cpp
clang-12: warning: Unknown CUDA version. cuda.h: CUDA_VERSION=11030. Assuming the latest supported version 10.1 [-Wunknown-cuda-version]
error: unable to load plugin '/mnt/lustre/users/wrs505/indv-project/sycl-workspace/hipSYCL/bin/../lib/libhipSYCL_clang.so': '/mnt/lustre/users/wrs505/indv-project/sycl-workspace/hipSYCL/bin/../lib/libhipSYCL_clang.so: undefined symbol: _ZN4llvm24DisableABIBreakingChecksE'
make: *** [main.o] Error 1
```
Below is the output from `clang -v`
```
clang version 12.0.1
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/apps/easybuild/software/Clang/12.0.1-GCCcore-11.2.0/bin
Found candidate GCC installation: /opt/apps/easybuild/software/GCCcore/11.2.0/lib/gcc/x86_64-pc-linux-gnu/11.2.0
Selected GCC installation: /opt/apps/easybuild/software/GCCcore/11.2.0/lib/gcc/x86_64-pc-linux-gnu/11.2.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
```
Any help is appreciated :)

**Comment by illuhad** [comment]
Can you please post `$HIPSYCL_INSTALL_PREFIX/etc/hipSYCL/syclcc.json`?

@sbalint98 I remember that we had issues with the ABI breaking checks with certain ROCm clang distributions. Do you remember what the cause was there?

**Comment by WageeshaR** [comment]
This is it.
```
{
  "version-major" : "0",
  "version-minor" : "9",
  "version-patch" : "2",
  "plugin-llvm-version-major" : "12",
  "plugin-with-cpu-acceleration" : "true",
  "default-clang"     : "/opt/apps/easybuild/software/Clang/12.0.1-GCCcore-11.2.0/bin/clang++",
  "default-nvcxx"     : "NVCXX_COMPILER-NOTFOUND",
  "default-platform"  : "cuda",
  "default-cuda-path" : "/opt/apps/easybuild/software/CUDA/11.3.1",
  "default-gpu-arch"  : "",
  "default-cpu-cxx"   : "/opt/apps/easybuild/software/GCCcore/11.2.0/bin/c++",
  "default-rocm-path" : "/opt/rocm",
  "default-use-bootstrap-mode" : "false",
  "default-is-dryrun" : "false",
  "default-use-accelerated-cpu" : "true",
  "default-clang-include-path" : "/opt/apps/easybuild/software/Clang/12.0.1-GCCcore-11.2.0/lib/clang/12.0.1/include/..",
  "default-sequential-link-line" : "-L/opt/apps/easybuild/software/Boost/1.77.0-GCC-11.2.0/lib -lboost_context -lboost_fiber -lomp -Wl,-rpath=/opt/apps/easybuild/software/Boost/1.77.0-GCC-11.2.0/lib",
  "default-sequential-cxx-flags" : "-I/opt/apps/easybuild/software/Boost/1.77.0-GCC-11.2.0/include -D_ENABLE_EXTENDED_ALIGNED_STORAGE",
  "default-omp-link-line" : "-L/opt/apps/easybuild/software/Boost/1.77.0-GCC-11.2.0/lib -lboost_context -lboost_fiber -Wl,-rpath=/opt/apps/easybuild/software/Boost/1.77.0-GCC-11.2.0/lib -fopenmp",
  "default-omp-cxx-flags" : "-I/opt/apps/easybuild/software/Boost/1.77.0-GCC-11.2.0/include -fopenmp -D_ENABLE_EXTENDED_ALIGNED_STORAGE",
  "default-rocm-link-line" : "-Wl,-rpath=$HIPSYCL_ROCM_PATH/lib -Wl,-rpath=$HIPSYCL_ROCM_PATH/hip/lib -L/opt/rocm/lib -L/opt/rocm/hip/lib -lamdhip64",
  "default-rocm-cxx-flags" : "-isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike -isystem /opt/apps/easybuild/software/Clang/12.0.1-GCCcore-11.2.0/lib/clang/12.0.1/include/.. -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I$HIPSYCL_ROCM_PATH/include -I$HIPSYCL_ROCM_PATH/include --rocm-device-lib-path=$HIPSYCL_ROCM_PATH/amdgcn/bitcode --rocm-path=$HIPSYCL_ROCM_PATH -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__",
  "default-cuda-link-line" : "-Wl,-rpath=$HIPSYCL_CUDA_LIB_PATH -L$HIPSYCL_CUDA_LIB_PATH -lcudart",
  "default-cuda-cxx-flags" : "-U__FLOAT128__ -U__SIZEOF_FLOAT128__ -isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike",
  "default-is-explicit-multipass" : "false",
  "default-save-temps" : "false"
}
```
Apparently it compiles for `--hipsycl-targets=omp.library-only`. But fails for all other passes.

**Comment by illuhad** [comment]
Thank you, that looks correct. For `omp.library-only`, it does not use the clang plugin, so that is why that one works..
There might be a mismatch between clang headers and binaries.
Do you know which compile flags easybuild has used to compile clang/LLVM?
It could also be that LLVM does not correctly advertise to cmake the libraries that it needs applications to link against. You can also try to add `LLVMSupport` to the list of linked libraries in `src/compiler/CMakeLists.txt` (line 69).

**Comment by WageeshaR** [comment]
I think there had been some conflict with modules I used to load LLVM/Clang ..etc. I used the `module restore` command to restore my previous snapshot which gave errors, however purging and reloading modules made it work. I can successfully compile for both `omp.library-only` and `omp.accelerated` modes. For CUDA, I don't get above error anymore although there are some CUDA device specific errors. Thanks for the support. Closing this now.
